docs/NetEbpfExtMultiAttach.md

```diff
@@ -0,0 +1,126 @@
+contexts. A client context represents the eBPF program that is attached to this specific "attach point". It also holds
+the attach params with which the program was attached. For each client context, a filter context is created. The filter
+context is container for all the WFP filters that are configured for this "attach point". Since only one client is
+allowed to attach for each attach parameter (i.e. "attach point"), there is a one-to-one mapping between client context
+and filter context.
+
+Whenever a new eBPF program is attached, following is the existing flow:
+1. Check the list of current client contexts in hook provider, to see if any other program with same attach params is
+already attached. This check happens with hook provider lock held. (There is a bug today in this logic where this lock
+is released after checking the list, causing 2 threads to both find that no matching client exists, and both threads
+can then proceed with attach for clients with same attach params).
+2. If a program is already attached, reject the new attach request.
+3. If no program is already attached with same attach params, create a new client context.
+4. Create a new filter context for this client context and add WFP filters.
+4. Add the new client context in the list of clients in the hook provider.
+
+## Synchronization
+
+There are 2 flows that need synchronization -- program invocation by the hook, and client attach and detach callbacks.
+In the current design, this synchronization is achieved by using rundown references.
+
+**Program invocation flow**:
+1. Acquire rundown reference for the client context. If this fails, bail.
+2. Invoke program.
+3. Release rundown reference for client context.
+
+**Client attach callback flow**
+1. Filter context that is created on program attach takes an initial rundown reference on the client context.
+2. Each WFP filter that is added takes a normal reference on filter context.
+
+**Client detach callback flow**
+1. Delete all the WFP filters. These will eventually release references on the filter context.
+2. Queue a work item that waits for rundown on client context. This blocks until all the current invocations of the
+program are completed. This also fails any new invocations of the program.
+3. When all the references to the filter context are released (as part of WFP filter delete notification), it releases
+the initial rundown reference taken on the client context, and the work item is unblocked, which then completes the NMR
+client detach callback.
+
+# New Design
+
+In the new design, multiple programs attaching to the same attach point (i.e. same attach params) is allowed. The
+required data structures changes are described below.
+
+Even though multiple programs will attach with same "attach params", there will be a single instance of the
+corresponding WFP filters added. This means there will be one-to-many mapping from filter context to client context.
```

Comment:
For the benefit of the reader please establish that the client to the hook is an eBPF program. So program and client are used interchangeably in this doc.